pandas/tests/reshape/test_merge.py

```diff
@@ -1515,6 +1515,33 @@ def test_self_join_multiple_categories(self):
 
         assert_frame_equal(result, df)
 
+    def test_categorical_dates(self):
+        # GH 16900
+        # dates should not be coerced to ints
+
+        df = pd.DataFrame(
+            [[date(2001, 1, 1), 1.1],
+             [date(2001, 1, 2), 1.3]],
+            columns=['date', 'num2']
+        )
+        df['date'] = df['date'].astype('category')
+
+        df2 = pd.DataFrame(
+            [[date(2001, 1, 1), 1.3],
+             [date(2001, 1, 3), 1.4]],
+            columns=['date', 'num4']
+        )
+        df2['date'] = df2['date'].astype('category')
+
+        result = pd.merge(df, df2, how='outer', on=['date'])
+        assert result.shape == (3, 3)
```

Comment:
construct the expected result and use ``tm.assert_frame_equal`` for both examples